File: python/maximum_subarray.py
from typing import List


class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        total = 0
        max_sum = nums[0]
        for num in nums:
            total += num
            max_sum = max(total, max_sum)
            if total < 0:
                total = 0

        return max_sum  
        

File: go/maximum_subarray/maximum_subarray.go
package kata

// ::KATA START::
func maxSubArray(nums []int) int {
	if len(nums) == 0 {
		return 0
	}

	subtotal := 0
	maximum := nums[0]
	for _, num := range nums {
		subtotal += num
		maximum = max(subtotal, maximum)
		if subtotal < 0 {
			subtotal = 0
		}
	}
	return maximum

	// brute force
	// currMax := nums[0]
	// for i, n := 0, len(nums); i < n; i++ {
	//    currMax = max(nums[i], currMax)
	//
	//    sum := nums[i]
	// 	for j := i + 1; j < n; j++ {
	//      sum += nums[j]
	//      currMax = max(sum, currMax)
	// 	}
	// }
	//  return currMax
}

// ::KATA END::
